internal/alb/sg/association.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	err := controller.sgController.Reconcile(lbSG)
+	if err != nil {
+		return lbSG, fmt.Errorf("failed to reconcile managed LoadBalancer securityGroup due to %s", err.Error())
+	}
+	lbSGAttachment := &LbAttachment{
+		GroupIDs: []string{*lbSG.GroupID},
+		LbArn:    association.LbArn,
+	}
+	err = controller.lbAttachmentController.Reconcile(lbSGAttachment)
+	if err != nil {
+		return lbSG, fmt.Errorf("failed to reconcile managed LoadBalancer securityGroup attachment due to %s", err.Error())
+	}
+	return lbSG, nil
+}
+
+func (controller *associationController) reconcileManagedInstanceSG(association *Association, lbSG *SecurityGroup) error {
+	instanceSGName := controller.namer.NameInstanceSG(association.LbID)
+	instanceSG := &SecurityGroup{
+		GroupName: &instanceSGName,
+		InboundPermissions: []*ec2.IpPermission{
+			{
+				IpProtocol: aws.String("tcp"),
+				FromPort:   aws.Int64(0),
+				ToPort:     aws.Int64(65535),
+				UserIdGroupPairs: []*ec2.UserIdGroupPair{
+					{
+						GroupId: lbSG.GroupID,
+					},
+				},
+			},
+		},
+	}
+	err := controller.sgController.Reconcile(instanceSG)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile managed Instance securityGroup due to %s", err.Error())
+	}
+	instanceSGAttachment := &InstanceAttachment{
+		GroupID: *instanceSG.GroupID,
+		Targets: association.Targets,
+	}
+	err = controller.instanceAttachmentController.Reconcile(instanceSGAttachment)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile managed Instance securityGroup attachment due to %s", err.Error())
+	}
+	return nil
+}
+
+func (controller *associationController) deletedManagedSGs(association *Association) error {
+	err := controller.deleteManagedInstanceSG(association)
+	if err != nil {
+		return fmt.Errorf("failed to delete managed Instance securityGroup due to %s", err.Error())
+	}
+
+	err = controller.deleteManagedLbSG(association)
+	if err != nil {
+		return fmt.Errorf("failed to delete managed LoadBalancer securityGroup due to %s", err.Error())
+	}
+	return nil
+}
+
+func (controller *associationController) deleteManagedLbSG(association *Association) error {
+	lbSGName := controller.namer.NameLbSG(association.LbID)
+	lbSGID, err := controller.findSGIDByName(lbSGName)
+	if err != nil {
+		return err
+	}
+	if lbSGID == nil {
+		return nil
+	}
+	lbSGAttachment := &LbAttachment{
+		GroupIDs: []string{*lbSGID},
+		LbArn:    association.LbArn,
+	}
+	err = controller.lbAttachmentController.Delete(lbSGAttachment)
+	if err != nil {
+		return err
+	}
+	lbSG := &SecurityGroup{
+		GroupID: lbSGID,
+	}
+	err = controller.sgController.Delete(lbSG)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (controller *associationController) deleteManagedInstanceSG(association *Association) error {
+	instanceSGName := controller.namer.NameInstanceSG(association.LbID)
+	instanceSGID, err := controller.findSGIDByName(instanceSGName)
+	if err != nil {
+		return err
+	}
+	if instanceSGID == nil {
```

Comment:
Could `instanceSGID` be empty string? And is this a valid case if it happens?

Reply:
it might be empty string, since it could be deleted manually from aws console, or during previous reconcile process(and the controller crashed or errored for some reason after that)

Reply:
`if aws.StringValue(instanceSGID) == ""` ?

Reply:
it's better to test for nil as it is, since it's model `optional` concept 😄